src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java

```diff
@@ -852,252 +520,275 @@ public ItemStack getTotalAmount(@Nullable final ItemStack stack)
-                        getInventory().setInventorySlotContents(slot, new ItemStack(Items.BUCKET));
-                    }
-                    else
-                    {
-                        getInventory().decrStackSize(slot, 1);
-                    }
+                    new InvWrapper(getInventory()).extractItem(slot, 1, false);
                     reduceNeededResources(tempStack);
                 }
             }
         }
 
-        if (Configurations.builderBuildBlockDelay > 0 && block != Blocks.AIR)
+        if (Configurations.builderBuildBlockDelay > 0 && blockToPlace != Blocks.AIR)
         {
             setDelay(Configurations.builderBuildBlockDelay);
         }
 
         return true;
     }
 
+    /**
+     * Handle flower pots and spawn the right flower in it.
+     * @param pos the position.
+     */
+    public void handleFlowerPots(@NotNull final BlockPos pos)
+    {
+        /**
+         * Should be overwritten and implemented by certain entity if required.
+         */
+    }
+
+    /**
+     * On placement of a Container execute this to store the location in the regarding building.
+     * @param pos the position of the container.
+     */
+    public void connectChestToBuildingIfNecessary(@NotNull final BlockPos pos)
+    {
+        /**
+         * Classes can overwrite this if necessary.
+         */
+    }
+
+    /**
+     * Reduces the needed resources by 1.
+     *
+     * @param stack the stack which has been used now.
+     */
+    public void reduceNeededResources(final ItemStack stack)
+    {
+        /**
+         * Nothing to be done here. Workers overwrite this if necessary.
+         */
+    }
+
+    /**
+     * Get the worker of the AI.
+     * @return the EntityCitizen object.
+     */
+    public EntityCitizen getWorker()
+    {
+        return this.worker;
+    }
+
+    /**
+     * Get the entity of an entityInfo object.
+     *
+     * @param entityInfo the input.
+     * @return the output object or null.
+     */
+    @Nullable
+    public Entity getEntityFromEntityInfoOrNull(Template.EntityInfo entityInfo)
+    {
+        try
+        {
+            return EntityList.createEntityFromNBT(entityInfo.entityData, world);
+        }
+        catch (RuntimeException e)
+        {
+            Log.getLogger().info("Couldn't restore entitiy", e);
+            return null;
+        }
+    }
+
+    /**
+     * Set the currentStructure to null.
+     */
+    public void resetCurrentStructure()
+    {
+        currentStructure = null;
+    }
+
+    /**
+     * Iterates through all entities and spawns them
+     * Suppressing Sonar Rule Squid:S3047
+     * The rule thinks we can merge the two forge loops iterating over resources
+     * But in this case the rule does not apply because that would destroy the logic.
+     */
+    @SuppressWarnings("squid:S3047")
     private Boolean spawnEntity(@NotNull final Structure.StructureBlock currentBlock)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 32 to the 20 allowed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)